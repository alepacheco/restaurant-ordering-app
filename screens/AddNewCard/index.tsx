import React, { useState, useEffect } from 'react';
import styled from 'styled-components/native';
import { TextInput } from 'components/Forms/TextInput';
import { Header } from 'components/Header';
import { savePaymentMethods } from 'utils/network';
import { useStoreActions } from 'store';
import { getPaymentMethods } from 'utils/network';
import { WideButton } from 'components/Button/WideButton';

const Wrapper = styled.SafeAreaView`
  background-color: ${props => props.theme.contrast0};
  height: 100%;
`;

const AddButtonWrapper = styled.TouchableOpacity`
  margin: 32px 16px;
  padding: 12px;
  background-color: ${props => props.theme.contrast2};
  border-radius: 6px;
`;

const AddButtonText = styled.Text`
  text-align: center;
  color: ${props => props.theme.contrast9};
`;

const Row = styled.View`
  display: flex;
  flex-direction: row;
`;

const Space = styled.View`
  height: 35px;
`;

const AddButton: React.FC<{ onPress: any }> = ({ onPress }) => {
  return (
    <AddButtonWrapper onPress={onPress}>
      <AddButtonText>Add card</AddButtonText>
    </AddButtonWrapper>
  );
};

export const AddNewCard: React.FC<{ navigation: any }> = ({ navigation }) => {
  const setPaymentMethods = useStoreActions(
    actions => actions.user.setPaymentMethods
  );

  const [cardNumber, setCardNumber] = useState();
  const [expMonth, setExpMonth] = useState();
  const [expYear, setExpiryYear] = useState();
  const [cvc, setCvc] = useState();

  return (
    <Wrapper>
      <Header title="Add new card" />
      <Space />
      <TextInput
        placeholder="Card Number"
        keyboardType="number-pad"
        autoCompleteType="cc-number"
        enablesReturnKeyAutomatically
        onChangeText={text => setCardNumber(text)}
        value={cardNumber}
        autoCapitalize="none"
        autoCorrect={false}
        blurOnSubmit
        maxLength={16}
      />

      <Row>
        <TextInput
          placeholder="Month"
          keyboardType="number-pad"
          autoCompleteType="cc-exp-month"
          enablesReturnKeyAutomatically
          onChangeText={text => setExpMonth(text)}
          value={expMonth}
          autoCapitalize="none"
          autoCorrect={false}
          blurOnSubmit
          maxLength={2}
        />

        <TextInput
          keyboardType="number-pad"
          autoCompleteType="cc-exp-year"
          enablesReturnKeyAutomatically
          placeholder="Year"
          onChangeText={text => setExpiryYear(text)}
          value={expYear}
          autoCapitalize="none"
          autoCorrect={false}
          blurOnSubmit
          maxLength={4}
        />

        <TextInput
          keyboardType="number-pad"
          autoCompleteType="cc-csc"
          enablesReturnKeyAutomatically
          placeholder="CVC"
          onChangeText={text => setCvc(text)}
          value={cvc}
          autoCapitalize="none"
          blurOnSubmit
          maxLength={4}
        />
      </Row>

      <WideButton
        text="Add card"
        type="secondary"
        onClick={() => {
          savePaymentMethods({
            number: cardNumber,
            expMonth,
            expYear,
            cvc,
          }).then(res => {
            navigation.goBack();
            getPaymentMethods().then(payments => setPaymentMethods(payments));
          });
        }}
      />
    </Wrapper>
  );
};
