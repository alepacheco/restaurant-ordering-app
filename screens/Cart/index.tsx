import React from 'react';
import { View } from 'react-native';
import styled from 'styled-components/native';
import { useStoreState } from 'store';
import { Selection } from './Selection';
import { createOrder } from 'utils/network';
import { CheckoutButton } from './CheckoutButton';
import { Restaurant, MenuItem } from 'types/restaurant';
import { Header } from './Header';

const StyledView = styled.View`
  display: flex;
  flex: 1;
  flex-direction: column;
  background-color: #eeeeee;
`;

const Footer = styled.SafeAreaView`
  display: flex;
  flex-direction: column;
  ${props => `background-color: ${props.theme.color};`}
`;

const ItemsScrollView = styled.ScrollView``;

const StyledText = styled.Text`
  ${props => `color: ${props.theme.textColor};`}
`;

const mapOfItemsFromRestaurant = (
  restaurant: Restaurant
): { [itemId: string]: MenuItem } => {
  const itemsForRestaurant = restaurant.menu.reduce(
    (acc, item) => [...acc, ...item.items],
    [] as MenuItem[]
  );

  return itemsForRestaurant.reduce(
    (acc, item) => ({
      ...acc,
      [item._id]: item,
    }),
    {} as { [itemId: string]: MenuItem }
  );
};

export const Cart: React.FC<{ navigation: any }> = ({ navigation }) => {
  const { restaurantId }: { restaurantId: string } = navigation.state.params;

  const cartSelection = useStoreState(state => state.cart.items[restaurantId]);
  const restaurantDetails = useStoreState(
    state => state.restaurantDetails.list[restaurantId]
  );

  const map = mapOfItemsFromRestaurant(restaurantDetails);

  const selectionEntries = cartSelection.map((item, index) => (
    <Selection key={index} selectionData={item} itemData={map[item.itemId]} />
  ));

  return (
    <StyledView>
      <Header />
      <ItemsScrollView>
        <View>{selectionEntries}</View>
        <StyledText>Total: XXX</StyledText>
      </ItemsScrollView>
      <Footer>
        <CheckoutButton restaurantId={restaurantId} />
      </Footer>
    </StyledView>
  );
};
