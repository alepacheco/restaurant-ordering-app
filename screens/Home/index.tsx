import React from 'react';
import { Button, View } from 'react-native';

export class Home extends React.Component<{ navigation: any }> {
  static navigationOptions = {
    title: 'Welcome',
  };
  render() {
    const { navigate } = this.props.navigation;
    console.log(this.props);
    return (
      <View>
        <Button title="Go Settings" onPress={() => navigate('Settings')} />
      </View>
    );
  }
}
