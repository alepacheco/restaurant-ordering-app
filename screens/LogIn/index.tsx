import * as SecureStore from 'expo-secure-store';
import React, { useEffect, useState } from 'react';
import { SafeAreaView, Text, Button } from 'react-native';
import { SESSION_ID_KEY } from '../../constants/session';
import { StackActions, NavigationActions } from 'react-navigation';
import axios from 'axios';
import { API_URL } from 'react-native-dotenv';

const loginNow = async ({ reload }) => {
  const { data } = await axios.post(`${API_URL}/login`, {
    username: 'admin',
    password: 'admin',
  });

  await SecureStore.setItemAsync(SESSION_ID_KEY, data.sessionId);

  // Force the LogIn component to rerender
  reload(Math.random());
};

export const LogIn: React.FC<{ navigation: any }> = ({ navigation }) => {
  const [, reload] = useState();

  useEffect(() => {
    SecureStore.getItemAsync(SESSION_ID_KEY).then(sessionId => {
      if (sessionId) {
        const resetAction = StackActions.reset({
          index: 0,
          actions: [NavigationActions.navigate({ routeName: 'Home' })],
        });
        navigation.dispatch(resetAction);
      }
    });
  });

  return (
    <SafeAreaView>
      <Text>LogIn to continue</Text>
      <Button title="Log in now" onPress={() => loginNow({ reload })} />
    </SafeAreaView>
  );
};
