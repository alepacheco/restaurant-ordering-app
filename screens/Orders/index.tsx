import React, { useState, useEffect } from 'react';
import { Text, FlatList } from 'react-native';
import styled from 'styled-components/native';
import { getUserOrders } from 'utils/network';
import { Order } from './Order';

type Unpacked<T> = T extends Promise<infer U> ? U : T;

const Wrapper = styled.View`
  height: 100%;
  background-color: ${props => props.theme.contrast0_5};
`;

const Header = styled.View`
  margin-top: 38px;
`;

const HeaderText = styled.Text`
  color: ${props => props.theme.textColor};
  margin: 8px;
  font-weight: bold;
`;

const OrderScroll = styled(FlatList)`
  background-color: ${props => props.theme.contrast0_5};
`;

export const Orders: React.FC<{}> = ({}) => {
  const [userOrders, setUserOrders] = useState(
    [] as Unpacked<ReturnType<typeof getUserOrders>>
  );
  useEffect(() => {
    getUserOrders().then(data => setUserOrders(data));
  }, []);

  return (
    <Wrapper>
      <Header>
        <HeaderText>Orders</HeaderText>
      </Header>
      <OrderScroll
        data={userOrders}
        keyExtractor={(item: any) => item._id}
        renderItem={({ item }: any) => <Order {...item}></Order>}
      />
    </Wrapper>
  );
};
