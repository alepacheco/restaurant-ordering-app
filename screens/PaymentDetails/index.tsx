import React, { useEffect } from 'react';
import styled from 'styled-components/native';
import { useStoreActions, useStoreState } from 'store';
import { getPaymentMethods } from 'utils/network';
import { Card } from './Card';

const Wrapper = styled.View`
  background-color: ${props => props.theme.contrast0_5};
  height: 100%;
  padding-top: 34px;
`;

const Title = styled.Text`
  color: ${props => props.theme.textColor};
  font-weight: bold;
  margin: 18px 0 12px 12px;
  font-size: 24px;
`;

const AddCardText = styled.Text`
  color: ${props => props.theme.textColor};
  font-weight: bold;
`;

const AddCardButton = () => {
  return <AddCardText>Add new card</AddCardText>;
};

export const PaymentDetails: React.FC<{}> = () => {
  const paymentMethods = useStoreState(state => state.user.user.paymentMethods);
  const setPaymentMethods = useStoreActions(
    actions => actions.user.setPaymentMethods
  );

  useEffect(() => {
    if (!paymentMethods) {
      getPaymentMethods().then(payments => setPaymentMethods(payments));
    }
  }, [paymentMethods, setPaymentMethods]);

  console.log({ paymentMethods });

  return (
    <Wrapper>
      <Title>Payment options</Title>
      {paymentMethods &&
        paymentMethods.map(paymentMethod => {
          return (
            <Card
              key={paymentMethod.card.fingerprint}
              last4={paymentMethod.card.last4}
              brand={paymentMethod.card.brand}
            />
          );
        })}
      <AddCardButton />
    </Wrapper>
  );
};
