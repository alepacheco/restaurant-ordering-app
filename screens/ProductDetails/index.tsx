import React, { useState, useEffect } from 'react';
import { View, Text, SafeAreaView, ScrollView } from 'react-native';
import styled from 'styled-components/native';
import { QuantityForm } from './QuantityForm';
import { AddToCart } from './AddToCart';
import { NavigationInjectedProps } from 'react-navigation';
import { MenuItem } from 'types/restaurant';
import { useStoreState, useStoreActions } from 'store';
const StyledView = styled.View`
  ${props => `background-color: ${props.theme.color};`}
`;

const ProductImage = styled.Image`
  height: 84px;
`;

const Name = styled.Text`
  ${props => `color: ${props.theme.textColor};`}

  font-size: 28px;
  text-align: center;
`;

const Description = styled.Text`
  font-size: 18px;
  color: gray;
`;

const Wrapper = styled.View`
  height: 100%;
  display: flex;
  flex-direction: column;
`;

export const ProductDetails: React.FC<{
  navigation: any;
}> = ({ navigation }) => {
  const {
    itemId,
    restaurantId,
  }: { itemId: string; restaurantId: string } = navigation.state.params;

  const itemDetails: MenuItem = useStoreState(state =>
    state.restaurantDetails.list[restaurantId].menu
      .reduce((acc, item) => acc.concat(item.items), [] as Array<MenuItem>)
      .find(menuItem => menuItem._id === itemId)
  );

  const cartSelection = useStoreState(state => state.cart.items[restaurantId]);
  const addToCart = useStoreActions(actions => actions.cart.add);

  console.log(cartSelection);
  const [quantity, setQuantity] = useState(1);

  return (
    <StyledView>
      <SafeAreaView>
        <Wrapper>
          <ScrollView>
            <ProductImage
              resizeMode="contain"
              source={{ uri: itemDetails.imageUrl }}
            />
            <Name>{itemDetails.name}</Name>
            <Description>{itemDetails.description}</Description>

            <QuantityForm onChange={setQuantity} initialValue={quantity} />
          </ScrollView>
          <AddToCart
            price={(Number(itemDetails.price) * quantity).toFixed(2)}
            onPress={() =>
              addToCart({
                restaurantId,
                items: {
                  amount: quantity,
                  itemId,
                  options: [],
                },
              })
            }
          />
        </Wrapper>
      </SafeAreaView>
    </StyledView>
  );
};
