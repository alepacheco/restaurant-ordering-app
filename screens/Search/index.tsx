import React from 'react';
import { SafeAreaView, Text } from 'react-native';

export class Search extends React.Component<{}> {
  render() {
    return (
      <SafeAreaView>
        <Text>Search</Text>
      </SafeAreaView>
    );
  }
}
