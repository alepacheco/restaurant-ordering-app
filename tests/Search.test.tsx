import React from 'react';
import renderer from 'react-test-renderer';

import { Search } from '../screens/Search';

describe('<Search />', () => {
  it('renders correctly', async () => {
    let tree: null | renderer.ReactTestRenderer = null;

    await renderer.act(async () => {
      tree = renderer.create(<Search />);
    });
    expect(tree.toJSON()).toMatchSnapshot();
  });
});
