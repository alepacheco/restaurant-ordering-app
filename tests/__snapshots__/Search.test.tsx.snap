// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Search /> renders correctly 1`] = `
<RCTSafeAreaView
  emulateUnlessSupported={true}
>
  <Text>
    Search
  </Text>
</RCTSafeAreaView>
`;
